File: Practica1/src/practica1/Weapon.java
package practica1;


class Weapon {
    private String name;
    private WeaponType type;
    private int uses;
    private static final float DEFAULT_POWER=1.0f;
    
    Weapon(String name, WeaponType type, int uses){
        this.name=name;
        this.type=type;
        this.uses=uses;
    }
    
    Weapon(Weapon w){
        this(w.name, w.type, w.uses);
    }
    
    public WeaponType getType(){
        return type;
    }
    
    public int getUses(){
        return uses;
    }
    
    public float power(){
        return type.getPower();
    }
    
    public float useIt(){
        float result;
        if(uses>0){
            uses--;
            result=power();
        }else result=DEFAULT_POWER;
        
        return result;
    }
    
    
}

File: PRACTICA1/ruby/Weapon.rb
#encoding: utf-8
#Esta clase representa a las armas de las que puede disponer una estación 
#espacial para potenciar su energía al disparar
require_relative "WeaponType" 

module Deepspace
	class Weapon
		@DEFAULT_POWER = 1.0
		@MIN_USES = 0
		
		def initialize(name, type, uses)
			@name = name
			@type = type
			@uses = uses
		end
		
		def self.newCopy(weapon)
			new(weapon.name, weapon.type, weapon.uses)
		end
		
		def name
			@name
		end
		
		def type
			@type
		end
		
		def uses
			@uses
		end
		
		def power
			type.power
		end
		
		def self.min_uses
			@MIN_USES
		end
			
		def self.default_power
			@DEFAULT_POWER
		end
		
		def useIt
			aux=0
			if uses > self.class.min_uses then
				@uses-=1
				aux=power
			else
				aux= self.class.default_power
			end
			
			return aux
		end
	end
end
